### PR TITLE
feat(frontend)!: migrate to typescript

### PR DESCRIPTION
**Pull request by gtangelo**
Migrated the frontend codebase to use typescript over javascript. Notable changes:
- well everything...
- Added type notations and updated relevant config for typescript (i.e. eslint rules)
- Restrcuture some folders around

### CODE REVIEW

**Comment by Peedee2002** [comment]
Sorry how many files changed??????

**Comment by gtangelo** [comment]
> Sorry how many files changed??????

dw its only 20........7 haha. But mostly renames of the entire FE files to tsx 💀

File: frontend/.eslintrc [hunk 1]
```diff
@@ -33,7 +39,6 @@
         "jsx-a11y/click-events-have-key-events": ["off"],
         "jsx-a11y/interactive-supports-focus": ["off"],
         "quotes": ["error", "double"],
-        "no-shadow": ["error", { "allow": ["_"]}],
```
Comment (Peedee2002) [inline frontend/.eslintrc:36]: 😢 i like this rule

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import {
 } from "react-router-dom";
 import { ThemeProvider } from "styled-components";
 import ErrorBoundary from "components/ErrorBoundary";
+import { RootState } from "config/store";
```
Comment (Peedee2002) [inline frontend/src/App.tsx]:
```suggestion
import type { RootState } from "config/store";
```

File: frontend/src/components/CourseListModal/CourseListModal.tsx [hunk 3]
```diff
@@ -1,10 +1,17 @@
-import React from "react";
+import React, { Dispatch, SetStateAction } from "react";
 import CourseBadge from "pages/ProgressionChecker/CourseBadge";
 import S from "./styles";
 
+type Props = {
+  title: string
+  courses: any
```
Comment (Peedee2002) [inline frontend/src/components/CourseListModal/CourseListModal.tsx]: 🤔

File: frontend/src/components/DraggableTab/DraggableTab.tsx [hunk 4]
```diff
@@ -4,16 +4,22 @@ import { useDispatch, useSelector } from "react-redux";
 import { CloseOutlined } from "@ant-design/icons";
 import { Button } from "antd";
 import { useTheme } from "styled-components";
+import { RootState } from "config/store";
```
Comment (Peedee2002) [inline frontend/src/components/DraggableTab/DraggableTab.tsx]:
```suggestion
import type { RootState } from "config/store";
```

File: frontend/package.json [hunk 5]
```diff
@@ -37,12 +38,13 @@
   "scripts": {
     "start": "vite --port 3000 --host",
     "start:local": "VITE_BACKEND_API_BASE_URL=https://circlesapi.csesoc.app VITE_ENV=dev vite --port 3000 --host",
-    "build": "vite build",
+    "build": "export NODE_OPTIONS=--max_old_space_size=4096; vite build",
```
Comment (gtangelo) [inline frontend/package.json:42]: reminder for @gtangelo, check if this should be include/or if there is another way around it

File: frontend/.eslintrc [hunk 6]
```diff
@@ -57,12 +62,17 @@
...
+        "@typescript-eslint/no-unused-vars": ["off"],
+        "import/no-named-as-default": ["off"], // vite stuff
+        "react/require-default-props": ["off"] // typescript stuff
     },
     "settings": {
         "import/resolver": {
           "node": {
-            "moduleDirectory": ["node_modules", "src/"]
+            "moduleDirectory": ["node_modules", "src/"],
+            "extensions": [".js", ".jsx", ".ts", ".tsx"]
```
Comment (gtangelo) [inline frontend/.eslintrc]: dont forget to clean up eslint file

File: frontend/package.json (hunk 5 above)
Comment (Peedee2002) [reply frontend/package.json:42]: Yea I've seen this before, we should make another card to remove it. Also, we might crash Wheatley, ask @lhjt

Comment (lhjt) [reply frontend/package.json:42]: No code is built on Wheatley, so you don't have to worry about resource consumption at build time 😄

File: frontend/.eslintrc (hunk 1 above)
Comment (zax-xyz) [reply frontend/.eslintrc:36]: i don't

Comment (Peedee2002) [reply frontend/.eslintrc:36]: smh

Comment (gtangelo) [reply frontend/.eslintrc:36]: yea oof had to remove this rule because it conflicts with typescript enums :(

File: frontend/README.md [hunk 7]
```diff
@@ -33,8 +33,6 @@ In `frontend.env`, add:
 
 > NOTE: The `VITE_BACKEND_API_BASE_URL` environment variable is the base url endpoint that the backend is running on. If the environment variable is not specified, the react application will default to using `http://localhost:8000/` as the base url when calling the API endpoint.
 
-For features that are still in progress, we have a boolean flag which enables these features to be shown while developing. To enable this, simply add `VITE_ENV=dev` to `frontend.env`.
```
Comment (Peedee2002) [inline frontend/README.md:36]: Why is this removed

Comment (gtangelo) [reply frontend/README.md:36]: We can check if the app is in dev (and enable in dev features) via one of vite's default env vars https://vitejs.dev/guide/env-and-mode.html#env-variables, rather than using our own custom env vars (just realised we could do this LOL)

File: frontend/.eslintrc (hunk 1 above)
Comment (gtangelo) [reply frontend/.eslintrc:36]: fyi, this should should be already included in the default eslitn config for the ts version

**Comment by gtangelo** [comment]
reminder for @gtangelo. Todo the following
- [x] Test userflow is not broken, get people to peer test this
- [x] Check if FE can be run on other people's dev environment. Include vscode settings folder
- [x] Test if we need to do a migration

File: frontend/.eslintrc [hunk 8]
```diff
@@ -57,12 +54,20 @@
...
         }],
         "simple-import-sort/exports": ["error"],
-        "import/no-named-as-default": ["off"] // vite stuff
+        "@typescript-eslint/no-explicit-any": ["error"],
+        "@typescript-eslint/no-floating-promises": ["off"],
+        "react/require-default-props": ["off"], // off since we have typescript instead
+        "@typescript-eslint/no-misused-promises": ["error", { "checksVoidReturn": false }], 
+        "@typescript-eslint/ban-ts-comment": ["off"], // should be removed? - it's here to skip type checking on some files,
+        "@typescript-eslint/dot-notation": ["off"], // disable force use of . annotation when accessing object fields
+        "react-hooks/exhaustive-deps": ["warn"], // Checks effect dependencies
```
Comment (Peedee2002) [inline frontend/.eslintrc]: This should be an error. There is no reason to ever ignore it. if you're getting spammed, you should use callbacks instead.

Comment (gtangelo) [reply frontend/.eslintrc]: Agree it should be an error. But this PR is getting to big for any refactoring changes LOL. This rule was not enabled before moving to typescript so just added it as a warning for now so we can fix it in a future ticket/PR.

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
+    "eslint.workingDirectories": [
+        "frontend"
+    ]
+}
```
Comment (Peedee2002) [inline .vscode/settings.json:13]:
```suggestion
    "python.testing.unittestEnabled": false,
    "python.testing.pytestEnabled": true,
    "python.testing.cwd": "${workspaceFolder}/backend",
    "files.exclude": {
        "**/__pycache__": true,
        "**/.hypothesis": true,
        "**/.pytest_cache": true
    }
}
```

File: frontend/.eslintrc [hunk 10]
```diff
@@ -57,12 +54,19 @@
             ]
         }],
         "simple-import-sort/exports": ["error"],
-        "import/no-named-as-default": ["off"] // vite stuff
+        "@typescript-eslint/no-explicit-any": ["error"],
+        "@typescript-eslint/no-floating-promises": ["off"],
+        "react/require-default-props": ["off"], // off since we have typescript instead
+        "@typescript-eslint/no-misused-promises": ["error", { "checksVoidReturn": false }], 
+        "@typescript-eslint/dot-notation": ["off"], // disable force use of . annotation when accessing object fields
```
Comment (Peedee2002) [inline frontend/.eslintrc:60]: doesnt this violate type safety?

File: frontend/src/config/constants.ts [hunk 11]
```diff
@@ -13,17 +13,18 @@ import { getMostRecentPastTerm } from "pages/TermPlanner/utils";
...
-export const lightPurple = "#efdbff";
+export const lightYellow = '#f9b01e';
+export const lightGrey = '#565652';
+export const darkGrey = '#323739';
+export const yellow = '#FAAD14';
+export const purple = '#9254de';
+export const lightPurple = '#efdbff';
 
-export const inDev = import.meta.env.VITE_ENV === "dev";
+export const inDev = import.meta.env.DEV;
```
Comment (Peedee2002) [inline frontend/src/config/constants.ts]: isnt there also `NODE_ENV` or something?

File: frontend/src/pages/CourseSelector/CourseSearchBar/CourseSearchBar.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        );
+        setCourses(
+          Object.keys(res.data).map((course) => ({
+            label: `${course}: ${res.data[course]}`,
+            value: course,
+          })),
+        );
+      } catch (err) {
+        // eslint-disable-next-line
+        console.log(err);
```
Comment (Peedee2002) [inline frontend/src/pages/CourseSelector/CourseSearchBar/CourseSearchBar.tsx]: if we have established this pattern, might as well:
```suggestion
        console.log("Error at CourseSearchBar", err);
```

**Review by Peedee2002 — APPROVED** [review]
it seems that in the grand scheme of things, not much actually changed :)
thats a good thing. Hopefully this really helps with the dev experience tho.
BEFORE YOU MERGE THO PLZ LOOK AT MY SUGGESTIONS

File: .vscode/settings.json (hunk 9 above)
Comment (lhjt) [reply .vscode/settings.json:13]: I also second this addition (will just have to make sure you add the missing `,` to the previous item)

File: frontend/.eslintrc (hunk 10 above)
Comment (lhjt) [reply frontend/.eslintrc:60]: What part of this violates type safety? IIRC when you access a field via the `[key]` syntax, typescript checks to ensure that the `key` is typeof keyof whatever object you're accessing.

File: frontend/src/config/store.ts [hunk 13]
```diff
@@ -1,17 +1,25 @@
-import { configureStore } from "@reduxjs/toolkit";
-import { combineReducers } from "redux";
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-nocheck
```
Comment (lhjt) [inline frontend/src/config/store.ts]: What's the reason you don't want to type-check this file? I'm unsure as to how the `MigrationManifest` works, but I see removing this line does cause compilation to fail.

File: frontend/.eslintrc (hunk 10 above)
Comment (Peedee2002) [reply frontend/.eslintrc:60]: Oh nice

**Review by lhjt — APPROVED** [review]
Nice work; can build. I am also fine with merging this (after taking a look at the comments we have still left around 😄)

File: frontend/src/pages/CourseSelector/CourseSearchBar/CourseSearchBar.tsx (hunk 12 above)
Comment (gtangelo) [reply frontend/src/pages/CourseSelector/CourseSearchBar/CourseSearchBar.tsx]: Ooops must've missed this. cheers for the pickup!

**Comment by lhjt** [comment]
Another thing that could be considered is the error handling pattern - using `console.error()` potentially instead of `console.log()`?

Will have to open a PR later to add sentry support as well (outside the scope of this one).

File: frontend/src/config/constants.ts (hunk 11 above)
Comment (gtangelo) [reply frontend/src/config/constants.ts]: used `import.meta.env.MODE` for env comparison instead

**Review by lhjt — APPROVED** [review]

File: frontend/src/config/store.ts (hunk 13 above)
Comment (gtangelo) [reply frontend/src/config/store.ts]: Oh yea, didn't want to deal with trying to create types with the old redux state using the newer redux state types as you need to omit/transform the type if there's a major change with the structure. Would not be worth spending the time to do so as we will be moving the persistance to the BE soon 🤞

**Comment by gtangelo** [comment]
> Another thing that could be considered is the error handling pattern - using `console.error()` potentially instead of `console.log()`?
> 
> Will have to open a PR later to add sentry support as well (outside the scope of this one).

Done! Cheers @lhjt for the review :DDDD
